docs/proposals/015-abstract-algebra-interface-hierarchy.md

```diff
@@ -0,0 +1,347 @@
+Slang, providing a foundation for principled generic mathematical
+operations.
+
+## Status
+
+Status: Design Review
+
+Implementation: N/A
+
+Author: Ellie Hermaszewska
+
+Reviewer: TBD
+
+## Background
+
+Currently, Slang lacks a formal way to express algebraic properties of types.
+While basic arithmetic operations are supported, there's no standard vocabulary
+to express a type's conformance to algebraic laws. This is particularly
+relevant for graphics programming where we frequently work with a variety of
+mathematical objects, each fulfilling different properties described by
+abstract algebra, for example:
+
+- Vector spaces
+- Quaternions
+- Dual numbers
+- Matrices
+- Complex numbers
+
+Additionally, many types support only a subset of arithmetic operations. For
+example:
+
+- Strings support concatenation (forming a monoid) but not subtraction
+- Colors support addition and scalar multiplication but not division
+- Matrices support multiplication but aren't always invertible
+- Non-empty containers support concatenation (forming a semigroup) but lack an
+  identity element
+
+A single catch-all "Arithmetic" interface would be insufficient as it would
+force types to implement operations that don't make mathematical sense
+
+A formal interface hierarchy allows us to:
+
+- Write generic code that works across multiple numeric types
+- Clearly document mathematical properties
+- Avoid forcing inappropriate operations on types
+
+## Related Work
+
+- Haskell: Extensive typeclass hierarchy (Semigroup, Monoid)
+  - Haskell does have a catch-all `Num` class, which is largely considered a
+    misfeature, and several third-party packages exist to rectify this with a
+    more nuanced abstract algebra hierarchy
+- PureScript: Similar to Haskell but with stricter adherence to category theory
+  principles
+- Rust: Traits for arithmetic operations (Add, Mul)
+- Scala: Abstract algebra typeclasses in libraries like Spire
+
+## Proposed Approach
+
+Implement the following interfaces, replacing in part `__BuiltinArithmeticType`
+
+```slang
+// Semigroup represents types with an associative addition operation
+// Examples: Non-empty arrays (concatenation), Colors (blending)
+interface Semigroup
+{
+    T operator+(T a, T b);  // Associative addition
+};
+
+// Monoid adds an identity element to Semigroup
+// Examples: float3(0), identity matrix, zero quaternion
+interface Monoid : Semigroup
+{
+    static T zero();  // Additive identity
+};
+
+// Group adds inverse elements to Monoid
+// Examples: Vectors under addition, Matrices under addition
+interface Group : Monoid
+{
+    T operator-();  // Additive inverse
+};
+
+// CommutativeGroup ensures addition order doesn't matter
+// Examples: Vectors, Colors (in linear space)
+interface CommutativeGroup : Group
+{
+    // Enforces a + b == b + a
+};
+
+// Semiring adds multiplication with identity to Monoid
+// Examples: Matrices, Quaternions
```

Comment:
Matrices and quaternions are full rings, aren't they? Of course that means they're also semirings but it's more interesting to list examples here that are not full rings. I guess the non-negative integers would be an example. (?)